src/gui_common/CustomDialog.cs

```diff
@@ -0,0 +1,494 @@
+            {
+                // End a dragging operation
+                dragType = DragType.None;
+            }
+        }
+
+        if (@event is InputEventMouseMotion mouseMotion)
+        {
+            if (dragType == DragType.None)
+            {
+                HandlePreviewDrag(mouseMotion);
+            }
+            else
+            {
+                HandleActiveDrag();
+            }
+        }
+    }
+
+    /// <summary>
+    ///   This is overriden so mouse position could take the titlebar into account due to it being drawn
+    ///   outside of the normal Control's rect bounds.
+    /// </summary>
+    public override bool HasPoint(Vector2 point)
+    {
+        var rect = new Rect2(Vector2.Zero, RectSize);
+
+        var titleBarHeight = GetConstant("custom_titlebar_height", "WindowDialog");
+
+        // Enlarge upwards for title bar
+        var adjustedRect = new Rect2(
+            new Vector2(rect.Position.x, rect.Position.y - titleBarHeight),
+            new Vector2(rect.Size.x, rect.Size.y + titleBarHeight));
+
+        // Inflate by the resizable border thickness
+        if (Resizable)
+        {
+            var scaleborderSize = GetConstant("custom_scaleborder_size", "WindowDialog");
+
+            adjustedRect = new Rect2(
+                new Vector2(adjustedRect.Position.x - scaleborderSize, adjustedRect.Position.y - scaleborderSize),
+                new Vector2(adjustedRect.Size.x + scaleborderSize * 2, adjustedRect.Size.y + scaleborderSize * 2));
+        }
+
+        return adjustedRect.HasPoint(point);
+    }
+
+    /// <summary>
+    ///   Overrides the minimum size to account for default elements (e.g title, close button, margin) rect size
+    ///   and for the other custom added contents on the window.
+    /// </summary>
+    public override Vector2 _GetMinimumSize()
+    {
+        var margin = GetConstant("custom_margin", "Dialogs");
+        var font = GetFont("custom_title_font", "WindowDialog");
+
+        var buttonWidth = closeButton?.GetCombinedMinimumSize().x;
+        var titleWidth = font.GetStringSize(TranslationServer.Translate(windowTitle)).x;
+        var buttonArea = buttonWidth + (buttonWidth / 2);
+
+        var contentSize = Vector2.Zero;
+
+        for (int i = 0; i < GetChildCount(); ++i)
+        {
+            var child = GetChildOrNull<Control>(i);
+
+            if (child == null || child == closeButton || child.IsSetAsToplevel())
+                continue;
+
+            var childMinSize = child.GetCombinedMinimumSize();
+
+            contentSize = new Vector2(
+                Mathf.Max(childMinSize.x, contentSize.x),
+                Mathf.Max(childMinSize.y, contentSize.y));
+        }
+
+        // Redecide whether the largest rect is the default elements' or the contents'
+        return new Vector2(Mathf.Max(2 * buttonArea.GetValueOrDefault() + titleWidth,
+            contentSize.x + margin * 2), contentSize.y + margin * 2);
+    }
+
+    public void ClosePopup()
```

Comment:
So currently this does nothing?
But if/when something is done here, it would be super awesome to automatically be able to trigger that from just Hide or `Visible = false` call

Reply:
This method still needs to be implemented as I haven't got around with animating popup closing. Overriding the `Visible` property won't hurt, I guess it's worth to try.